### ISSUE TITLE
Potentially unnecessary copying performed during "no-op" tuple conversion with member count > 4

### ISSUE THREAD

**Issue by jrevels** [JuliaLang/julia#11765]
@mlubin @scidom I think I found the source of the unexpected tuple inefficiencies I emailed you about:

Conversion from an instance to that instance's type is generally a no-op:

```julia
julia> t4 = (1,2,3,4); T4 = typeof(t4);

julia> @code_llvm convert(T4,t4)

define [4 x i64] @julia_convert_21427(%jl_value_t*, [4 x i64]) {
top:
  ret [4 x i64] %1
}
```

I would expect this to be the case for tuples with greater than 4 members, as well, but that's not the case:

```julia
julia> t5 = (1,2,3,4,5); T5 = typeof(t5);

julia> @code_llvm convert(T5,t5)

define [5 x i64] @julia_convert_21498(%jl_value_t*, [5 x i64]) {
top:
  %const = bitcast i64 4583880832 to i64
  %2 = alloca [6 x %jl_value_t*], align 8
  %.sub3 = bitcast [6 x %jl_value_t*]* %2 to %jl_value_t**
  %3 = getelementptr [6 x %jl_value_t*]* %2, i64 0, i64 2
  %4 = bitcast [6 x %jl_value_t*]* %2 to i64*
  store i64 8, i64* %4, align 8
  %5 = load %jl_value_t*** @jl_pgcstack, align 8
  %6 = getelementptr [6 x %jl_value_t*]* %2, i64 0, i64 1
  %7 = bitcast %jl_value_t** %6 to %jl_value_t***
  store %jl_value_t** %5, %jl_value_t*** %7, align 8
  store %jl_value_t** %.sub3, %jl_value_t*** @jl_pgcstack, align 8
  store %jl_value_t* null, %jl_value_t** %3, align 8
  %8 = getelementptr [6 x %jl_value_t*]* %2, i64 0, i64 3
  store %jl_value_t* null, %jl_value_t** %8, align 8
  %9 = getelementptr [6 x %jl_value_t*]* %2, i64 0, i64 4
  store %jl_value_t* null, %jl_value_t** %9, align 8
  %10 = getelementptr [6 x %jl_value_t*]* %2, i64 0, i64 5
  store %jl_value_t* null, %jl_value_t** %10, align 8
  %const_mat1 = add i64 %const, 7424952
  %11 = inttoptr i64 %const_mat1 to %jl_value_t**
  %12 = load %jl_value_t** %11, align 8
  %13 = icmp eq %jl_value_t* %12, null
  br i1 %13, label %err, label %ok

err:                                              ; preds = %top
  call void @jl_undefined_var_error(%jl_value_t* inttoptr (i64 13183033256 to %jl_value_t*))
  unreachable

ok:                                               ; preds = %top
  %sunkaddr = ptrtoint [6 x %jl_value_t*]* %2 to i64
  %sunkaddr4 = add i64 %sunkaddr, 16
  %sunkaddr5 = inttoptr i64 %sunkaddr4 to %jl_value_t**
  store %jl_value_t* %12, %jl_value_t** %sunkaddr5, align 8
  %const_mat2 = add i64 %const, 26415400
  %14 = inttoptr i64 %const_mat2 to %jl_value_t**
  %15 = load %jl_value_t** %14, align 8
  %sunkaddr6 = ptrtoint [6 x %jl_value_t*]* %2 to i64
  %sunkaddr7 = add i64 %sunkaddr6, 24
  %sunkaddr8 = inttoptr i64 %sunkaddr7 to %jl_value_t**
  store %jl_value_t* %15, %jl_value_t** %sunkaddr8, align 8
  %16 = extractvalue [5 x i64] %1, 0
  %17 = insertvalue [1 x i64] undef, i64 %16, 0
  %18 = call %jl_value_t* @allocobj(i64 8)
  %19 = getelementptr inbounds %jl_value_t* %18, i64 -1, i32 0
  %const_mat = add i64 %const, -7117712
  %20 = inttoptr i64 %const_mat to %jl_value_t*
  store %jl_value_t* %20, %jl_value_t** %19, align 8
  %21 = bitcast %jl_value_t* %18 to [1 x i64]*
  store [1 x i64] %17, [1 x i64]* %21, align 8
  %sunkaddr9 = ptrtoint [6 x %jl_value_t*]* %2 to i64
  %sunkaddr10 = add i64 %sunkaddr9, 32
  %sunkaddr11 = inttoptr i64 %sunkaddr10 to %jl_value_t**
  store %jl_value_t* %18, %jl_value_t** %sunkaddr11, align 8
  %22 = extractvalue [5 x i64] %1, 1
  %23 = insertvalue [4 x i64] undef, i64 %22, 0
  %24 = extractvalue [5 x i64] %1, 2
  %25 = insertvalue [4 x i64] %23, i64 %24, 1
  %26 = extractvalue [5 x i64] %1, 3
  %27 = insertvalue [4 x i64] %25, i64 %26, 2
  %28 = extractvalue [5 x i64] %1, 4
  %29 = insertvalue [4 x i64] %27, i64 %28, 3
  %30 = inttoptr i64 %const to %jl_value_t*
  %31 = call [4 x i64] @julia_convert_21427(%jl_value_t* %30, [4 x i64] %29)
  %32 = call %jl_value_t* @allocobj(i64 32)
  %33 = getelementptr inbounds %jl_value_t* %32, i64 -1, i32 0
  %34 = inttoptr i64 %const to %jl_value_t*
  store %jl_value_t* %34, %jl_value_t** %33, align 8
  %35 = bitcast %jl_value_t* %32 to [4 x i64]*
  store [4 x i64] %31, [4 x i64]* %35, align 8
  %sunkaddr12 = ptrtoint [6 x %jl_value_t*]* %2 to i64
  %sunkaddr13 = add i64 %sunkaddr12, 40
  %sunkaddr14 = inttoptr i64 %sunkaddr13 to %jl_value_t**
  store %jl_value_t* %32, %jl_value_t** %sunkaddr14, align 8
  %36 = call %jl_value_t* @jl_f_apply(%jl_value_t* null, %jl_value_t** %3, i32 4)
  %37 = bitcast %jl_value_t* %36 to [5 x i64]*
  %38 = load [5 x i64]* %37, align 8
  %sunkaddr15 = ptrtoint [6 x %jl_value_t*]* %2 to i64
  %sunkaddr16 = add i64 %sunkaddr15, 8
  %sunkaddr17 = inttoptr i64 %sunkaddr16 to %jl_value_t***
  %39 = load %jl_value_t*** %sunkaddr17, align 8
  store %jl_value_t** %39, %jl_value_t*** @jl_pgcstack, align 8
  ret [5 x i64] %38
}
```

Both of these `convert` calls specialize to the same method:

```julia
julia> @which convert(T4,t4)
convert{T<:Tuple{Any,Vararg{Any}}}(::Type{T<:Tuple{Any,Vararg{Any}}}, x::Tuple{Any,Vararg{Any}}) at essentials.jl:96

julia> @which convert(T5,t5)
convert{T<:Tuple{Any,Vararg{Any}}}(::Type{T<:Tuple{Any,Vararg{Any}}}, x::Tuple{Any,Vararg{Any}}) at essentials.jl:96
```

The code for which is [here](https://github.com/JuliaLang/julia/blob/master/base/essentials.jl#L96).

**Comment by mlubin** [comment]
Does the issue go away if we add a definition:
```
convert{T<:Tuple{Any,Vararg{Any}}}(::Type{T}, x::T) = x
```

**Comment by jrevels** [comment]
Looks like it:

```julia
julia> Base.convert{T<:Tuple{Any,Vararg{Any}}}(::Type{T}, x::T) = x
convert (generic function with 517 methods)

julia> t = (1,2,3,4,5); T = typeof(t);

julia> @code_llvm convert(T,t)

define [5 x i64] @julia_convert_21439(%jl_value_t*, [5 x i64]) {
top:
  ret [5 x i64] %1
}
```

I'd be willing to make the PR to add this method from a fork, or it might be easier for you to do so since you have Collaborator status.

**Comment by mlubin** [comment]
I won't be committing this without a PR, so feel free to do the honors.

**Comment by JeffBezanson** [comment]
I would like to have such a definition but the last time I tried it had ambiguity problems. Maybe we can live with that though.

**Comment by mlubin** [comment]
@JeffBezanson, the new method is strictly more specific than an existing method, naively it seems like that shouldn't create any ambiguities.

**Comment by carnaval** [comment]
Fixed by #11767
